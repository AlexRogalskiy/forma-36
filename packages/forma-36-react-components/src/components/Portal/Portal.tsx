import { useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';

export interface PortalProps {
  children: React.ReactElement;
  container?: Document | HTMLElement;
}

export default function Portal({
  children,
  container: containerElement,
}: PortalProps): React.ReactPortal | null {
  const container = useRef<Document | HTMLElement | undefined>(
    containerElement,
  );
  const portal = useRef<HTMLDivElement>(document.createElement('div'));

  useEffect(() => {
    if (!container.current) {
      container.current = document.body;
    }

    container.current.appendChild(portal.current);

    return () => {
      if (container.current) {
        container.current.removeChild(portal.current);
      }
    };
  }, []);

  return portal.current ? createPortal(children, portal.current) : null;
}
