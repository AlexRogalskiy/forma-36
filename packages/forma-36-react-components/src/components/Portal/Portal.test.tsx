/* eslint-disable jsx-a11y/accessible-emoji */
import React from 'react';
import { shallow } from 'enzyme';

import Portal from './Portal';

it('renders the component', () => {
  const output = shallow(
    <Portal>
      <React.Fragment>ðŸ‘‹</React.Fragment>
    </Portal>,
  );
  expect(output).toMatchSnapshot();
});

it('renders the component in a separate container', () => {
  const container = document.createElement('span');
  document.body.appendChild(container);

  const output = shallow(
    <Portal container={container}>
      <React.Fragment>ðŸ‘‹</React.Fragment>
    </Portal>,
  );
  expect(output).toMatchSnapshot();
});
