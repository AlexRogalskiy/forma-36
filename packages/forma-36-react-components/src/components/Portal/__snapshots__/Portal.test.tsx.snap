// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<Portal
  containerInfo={<div />}
>
  ğŸ‘‹
</Portal>
`;

exports[`renders the component in a separate container 1`] = `
<Portal
  containerInfo={<div />}
>
  ğŸ‘‹
</Portal>
`;
